File: spiral_traversal.java
import java.util.Scanner;
public class spiral_traversal {
    public static void print(int arr[][] ,int m,int n){
        int start_row=0;
        int start_col=0;
        int i;
        while(start_row<m && start_col<n){
            for(i=start_col;i<n;i++ ){
                System.out.print(arr[start_row][i]+", ");
            }
            start_row++;
            for(i=start_row;i<m;i++)System.out.print(arr[i][n-1]+", ");
            n--;
            if(start_row<m){
                for(i=n-1;i>=start_col;i--)System.out.print(arr[m-1][i]+", ");
                m--;
            }
            if(start_col<n){
                for(i=m-1;i>=start_row;i--)System.out.print(arr[i][start_col]+", ");
                start_col++;
            }
        }
        System.out.print("END");
        
    }
    public static void main(String[] args){
        Scanner sc=new Scanner(System.in);
        int m=sc.nextInt();
        int n=sc.nextInt();
        int arr[][]=new int[m][n];
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++)arr[i][j]=sc.nextInt();
        }
        
        print(arr,m,n);

    }
    
}

File: spiral_traversal.cpp
#include<bits/stdc++.h>
using namespace std;

void print(int arr[][4],int n,int m){
    int start_col=0;
    int start_row=0;
    int end_col=m-1;
    int end_row=n-1;
    int total=n*m;
    int x=0;
    while(x<total){
        
        for(int cc=start_col;cc<=end_col && x<total;cc++){
            cout<<arr[start_row][cc]<<" ";
            x++;
        }
        start_row++;
        for(int rr=start_row;rr<=end_row && x<total;rr++){
            cout<<arr[rr][end_col]<<" ";
            x++;
        }
        end_col--;
        for(int cc=end_col;cc>=start_col && x<total;cc--){
            cout<<arr[end_row][cc]<<" ";
            x++;
        }
        end_row--;
        for(int ss=end_row;ss>=start_row && x<total;ss--){
            cout<<arr[ss][start_col]<<" ";
            x++;
        }
        start_col++;
        
    }
}
int main(){
    int n,m;
    cin>>n>>m;
    int arr[n][4];
    for(int i=0;i<n;i++){
        for(int j=0;j<4;j++)cin>>arr[i][j];
    }
    print(arr,n,4);
}
